runtime/src/iree/base/internal/cpu.c

```diff
@@ -10,9 +10,144 @@
 #include "iree/base/internal/cpu.h"
 
 #include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
 
 //===----------------------------------------------------------------------===//
-// iree_cpu_*
+// Platform-specific processor data queries
+//===----------------------------------------------------------------------===//
+
+#if defined(IREE_PLATFORM_ANDROID) || defined(IREE_PLATFORM_LINUX)
+
+// NOTE: not all kernel versions have all of the cap bits we need defined so as
+// a practice we always define the feature bits we need locally.
+#include <sys/auxv.h>
+
+// OR's |field_bit| into |field_value| if |hwcap_bit| is set in |hwcap_value|.
+#define IREE_SET_IF_HWCAP(hwcap_value, hwcap_bit, field_value, field_bit)   \
+  (field_value) = (field_value) | iree_all_bits_set(hwcap_value, hwcap_bit) \
+                      ? field_bit                                           \
+                      : 0
+
+#if defined(IREE_ARCH_ARM_64)
+// https://docs.kernel.org/arm64/elf_hwcaps.html
+#define IREE_HWCAP_ASIMD (1 << 1)
+#define IREE_HWCAP2_I8MM (1 << 13)
+static void iree_cpu_query_data_arch_hwcaps(uint32_t hwcap, uint32_t hwcap2,
+                                            uint64_t* out_fields) {
+  IREE_SET_IF_HWCAP(hwcap, IREE_HWCAP_ASIMD, out_fields[0],
+                    IREE_CPU_DATA_FIELD_0_AARCH64_HAVE_ASIMDDP);
+  IREE_SET_IF_HWCAP(hwcap2, IREE_HWCAP2_I8MM, out_fields[0],
+                    IREE_CPU_DATA_FIELD_0_AARCH64_HAVE_I8MM);
+}
+#else
+static void iree_cpu_query_data_arch_hwcaps(uint32_t hwcap, uint32_t hwcap2,
+                                            uint64_t* out_fields) {
+  // Not supported on this arch.
+}
+#endif  // IREE_ARCH_*
+
+#undef IREE_SET_IF_HWCAP
+
+static void iree_cpu_initialize_from_platform(iree_allocator_t temp_allocator,
+                                              uint64_t* out_fields) {
+  uint32_t hwcap = getauxval(AT_HWCAP);
+  uint32_t hwcap2 = getauxval(AT_HWCAP2);
+  iree_cpu_query_data_arch_hwcaps(hwcap, hwcap2, out_fields);
+}
+
+#else
+
+static void iree_cpu_initialize_from_platform(iree_allocator_t temp_allocator,
+                                              uint64_t* out_fields) {
+  // No implementation available. CPU data will be all zeros.
+}
+
+#endif  // IREE_PLATFORM_*
+
+//===----------------------------------------------------------------------===//
+// Architecture-specific string lookup
+//===----------------------------------------------------------------------===//
+
+#define IREE_TEST_FIELD_BIT(field_key, field_value, bit_value)          \
+  if (iree_string_view_equal(key, IREE_SV(field_key))) {                \
+    *out_value = iree_all_bits_set((field_value), (bit_value)) ? 1 : 0; \
+    return true;                                                        \
+  }
+
+#if defined(IREE_ARCH_ARM_64)
+
+static bool iree_cpu_lookup_data_by_key_for_arch(
```

Comment:
While the ability to look up by string key is potentially useful, for my immediate need in builtins I'm better off looking up by symbolic constant.

Reply:
I can see that it's useful in `iree_hal_inline_module_device_query_i64`, but that's not my primary use case, since I want to query cpu info from C code.